pkg/routeagent_driver/handlers/ovn/connection.go

```diff
@@ -186,45 +188,85 @@ func getTLSConfig(k8sClientset clientset.Interface) (*tls.Config, error) {
 	return tlsConfig, nil
 }
 
-func discoverOvnKubernetesNetwork(ctx context.Context, k8sClientset clientset.Interface, zoneName string) (string, error) {
-	ovnDBPod, err := FindPod(ctx, k8sClientset, "name=ovnkube-db")
+func discoverOvnKubernetesNetwork(ctx context.Context, k8sClientset clientset.Interface,
+	dynamicClient dynamic.Interface, zoneName string,
+) (string, error) {
+	openshiftNetwork, err := FindOpenshiftNetwork(ctx, dynamicClient)
 	if err != nil {
 		return "", err
 	}
 
-	var nbdbAddress string
+	if openshiftNetwork != nil {
+		return discoverOpenshiftOvnKubernetesNetwork(ctx, k8sClientset)
+	}
+
+	return discoverKindOvnKubernetesNetwork(ctx, k8sClientset, zoneName)
+}
+
+/*
+
+openshift non-ic: in this case there will get the "app=ovnkube-node" and check for ovnkube-db service,
+if present we will the default Openshift db path
+
+openshift ic with one node per zone: if above service is not present we will return default Openshift socket path
+*/
+
+func discoverOpenshiftOvnKubernetesNetwork(ctx context.Context, k8sClientSet clientset.Interface) (string, error) {
+	ovnPod, err := FindPod(ctx, k8sClientSet, "app=ovnkube-node")
+	if err != nil {
+		return "", errors.Wrap(err, "error finding a pod with label \"app=ovnkube-node\"")
+	}
+
+	_, err = k8sClientSet.CoreV1().Services(ovnPod.Namespace).Get(ctx, ovnKubeService, metav1.GetOptions{})
+	if err == nil {
+		return defaultOpenshiftOVNNBDB, nil
+	}
+
+	return defaultOVNOpenshiftUnixSocket, nil
+}
+
+/*
+   The discovery method is different for each kind of deployment.
+        kind non-ic: Get the db pod with label name=ovnkube-db, if present will get the protocol and other details from it
+
+        kind ic with multiple node per zone: we parse every endpoint in the namespace ovn uses and find the one that has
+        same zone as the node in which route-agent runs. To get the ovn namespace we use pods running on app=ovnkube-node.
+
+        kind ic with one node per zone: we will have no endpoints matching hence we will return default kind socket path
+*/
+
+func discoverKindOvnKubernetesNetwork(ctx context.Context, k8sClientSet clientset.Interface, zoneName string) (string, error) {
+	ovnDBPod, err := FindPod(ctx, k8sClientSet, "name=ovnkube-db")
+	if err != nil {
+		return "", err
+	}
 
 	if ovnDBPod != nil {
-		nbdbAddress, err = discoverOvnDBClusterNetwork(ctx, k8sClientset, ovnDBPod)
-	} else {
-		nbdbAddress, err = discoverOvnNodeClusterNetwork(ctx, k8sClientset, zoneName)
+		return discoverKindOvnDBClusterNetwork(ctx, ovnDBPod, k8sClientSet)
 	}
 
+	ovnPod, err := FindPod(ctx, k8sClientSet, "app=ovnkube-node")
```

Comment:
Looks like you are using ovnPod just for the namespace. Can we use the namespace of ovnDBPod? or do you think they can be deployed in different namespaces?

Reply:
This if for the case where we are not able to find a db pod.